weather_mv/loader_pipeline/regrid.py

```diff
@@ -32,16 +36,148 @@
+    """Regrid a xarray.Dataset with MetView.
+
+    Attributes:
+        regrid_kwargs: A dictionary of keyword-args to be passed into `mv.regrid()`
+            (excluding the dataset).
+        zarr_input_chunks: (Optional) When regridding Zarr data, how the input
+            dataset should be chunked upon open.
+    """
+    regrid_kwargs: t.Dict
+    zarr_input_chunks: t.Optional[t.Dict] = None
+
+    def template(self, source_ds: xr.Dataset) -> xr.Dataset:
+        """Calculate the output Zarr template by regridding a tiny slice of the input dataset."""
+        # Silence Dask warning...
+        with dask.config.set(**{'array.slicing.split_large_chunks': False}):
+            zeros = source_ds.chunk().pipe(xr.zeros_like)
+
+            # Get a single timeslice of the zeros Dataset (or equivalent chunkable dimension).
+            # We don't know for sure that 'time' is in the Zarr dataset, so here we make our
+            # best attempt to find a good slice.
+            t0 = None
+            for dim in ['time', *(self.zarr_input_chunks or {}).keys()]:
+                if dim in zeros:
+                    t0 = zeros.isel({dim: 0}, drop=True)
+                    break
+            if t0 is None:
+                raise ValueError('cannot infer any dimension when creating a Zarr template. '
+                                 'Please define at least one chunk in `--zarr_input_chunks`.')
+
+            _, ds = self._apply(xbeam.Key(), t0)
+            # Regrid the single time, then expand the Dataset to span all times.
+            tmpl = (
+                ds
+                .chunk()
+                .expand_dims({dim: zeros[dim]}, 0)
+            )
+
+            return tmpl
+
+    def apply(self, key: xbeam.Key, fs: Fieldset) -> t.Tuple[xbeam.Key, Fieldset]:
+        return key, mv.regrid(data=fs, **self.regrid_kwargs)
 
 
 @dataclasses.dataclass
 class Regrid(ToDataSink):
+    """Regrid data using MetView.
+
+    See https://metview.readthedocs.io/en/latest/metview/using_metview/regrid_intro.html
+    for an in-depth intro on regridding with MetView.
+
+    Attributes:
+        output_path: URI for regridding target. Can be a glob pattern of NetCDF or Grib files; optionally,
+            it can be a Zarr corpus is supported.
+        regrid_kwargs: A dictionary of keyword-args to be passed into `mv.regrid()` (excluding the dataset).
+        to_netcdf: When set, it raw data output will be written as NetCDF. Cannot use with Zarr datasets.
+        zarr_input_chunks: (Optional) When regridding Zarr data, how the input dataset should be chunked upon open.
+        zarr_output_chunks: (Optional, recommended) When regridding Zarr data, how the output Zarr dataset should be
+            divided into chunks.
+    """
     output_path: str
     regrid_kwargs: t.Dict
     to_netcdf: bool = False
+    zarr_input_chunks: t.Optional[t.Dict] = None
+    zarr_output_chunks: t.Optional[t.Dict] = None
+
+    def __post_init__(self):
```

Comment:
The fact that this is here makes me wish this class wasn't a dataclass anymore. I feel like anytime you're actually trying to manipulate the things stored in the class attributes as they're getting set, you should probably just write the init method.

Reply:
Changed some default so that this post-init is no longer necessary :)